File: Reverse Nodes in k-Group/Reverse_Nodes_in_k_Group.py
# 56ms 96.14%
class Solution:
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head:
            return None
        
        mid = head
        stack, flag, last_node = [], True, ListNode(0)
        
        while True:
            for i in range(k):
                stack.append(mid)
                mid = mid.next
                if mid is None and i < k - 1:
                    return head
            
            reverse_head = reverse_mid = stack.pop()
            
            if flag:
                head = reverse_mid
                flag = False
            
            while stack:    
                reverse_mid.next = stack.pop()
                reverse_mid = reverse_mid.next
            
            reverse_mid.next = mid
            last_node.next = reverse_head
            last_node = reverse_mid

            if mid is None:
                return head

File: Reverse Nodes in k-Group/Reverse_Nodes_in_k-Group.cpp
// 其实并不难，就是比较麻烦，如果非要说技巧，应该是单链表翻转的时候如何把头结点的next指针置空
// Runtime: 16 ms, faster than 96.91% of C++ online submissions for Reverse Nodes in k-Group.
// Memory Usage: 9.9 MB, less than 35.48% of C++ online submissions for Reverse Nodes in k-Group.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution 
{
public:
    ListNode* reverseKGroup(ListNode* head, int k) 
    {   
        // 首先统计一下链表的长度
        int length = 0;
        ListNode* temp = head;
        while (temp)
        {
            ++length;
            temp = temp->next;
        }
        
        // 边界条件处理
        if (head == nullptr || k <= 0 || k > length) return head;
        
        ListNode dummyHead(666);
        ListNode* preHead = &dummyHead;
        
        int numOfLeft = length;
        ListNode *preNode = nullptr, *curNode = head, *nextNode = nullptr;
        
        while (numOfLeft)
        {
            if (numOfLeft < k) break;
            
            ListNode *curHead = curNode, *curTail = nullptr;
            
            for (int i = 0; i < k; ++i)
            {   
                nextNode = curNode->next;
                curNode->next = preNode;
                
                preNode = curNode;
                curNode = nextNode;
                
                --numOfLeft;
            }

            curTail = preNode;
            
            preHead->next = curTail;
            curHead->next = curNode;
            
            preHead = curHead;
        }
        
        return dummyHead.next;
    }
};